File: core/src/main/java/org/predicode/predicator/terms/Phrase.java
package org.predicode.predicator.terms;

import org.predicode.predicator.Knowns;
import org.predicode.predicator.Rule;
import org.predicode.predicator.grammar.TermPrinter;
import org.predicode.predicator.predicates.Predicate;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import javax.annotation.Nonnull;
import javax.annotation.concurrent.Immutable;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.predicode.predicator.grammar.TermPrinter.printTerms;
import static org.predicode.predicator.terms.Variable.temp;


/**
 * A phrase consisting of other terms.
 */
@Immutable
public class Phrase extends CompoundTerm implements Predicate {

    @Nonnull
    private final List<? extends Term> terms;

    /**
     * Constructs phrase.
     *
     * @param terms list of terms this phrase consists of.
     */
    public Phrase(@Nonnull List<? extends Term> terms) {
        this.terms = Collections.unmodifiableList(terms);
    }

    /**
     * Constructs phrase out of terms array.
     *
     * @param terms array of terms this phrase consists of.
     */
    public Phrase(@Nonnull Term ...terms) {
        this(Arrays.asList(terms));
    }

    /**
     * List of terms this phrase consists of.
     *
     * @return readonly list.
     */
    @Nonnull
    public final List<? extends Term> getTerms() {
        return this.terms;
    }

    @Nonnull
    @Override
    public Mono<Expansion> expand(@Nonnull Resolver resolver) {
        return expansion(resolver)
                .map(phraseExpansion -> {

                    final Variable tempVar = temp("phrase expansion");

                    return phraseExpansion.resolver.getKnowns().declareLocal(
                            tempVar,
                            (local, knowns) -> new Expansion(
                                    tempVar,
                                    knowns,
                                    predicate -> phraseExpansion.definition(local).and(predicate)));
                })
                .next();
    }

    /**
     * Resolves this phrase as predicate.
     *
     * <p>{@link Term#expand(Resolver) Expands} all of the phrase terms, then searches for corresponding
     * {@link Rule resolution rules} and applies them.</p>
     */
    @Override
    @Nonnull
    public Flux<Knowns> resolve(@Nonnull Resolver resolver) {
        try {
            return expansion(resolver)
                    .flatMap(PhraseExpansion::resolve);
        } catch (Exception e) {
            return Flux.error(e);
        }
    }

    @Override
    public void print(@Nonnull TermPrinter out) {
        out.startCompound();
        out.print(getTerms());
        out.endCompound();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        final Phrase phrase = (Phrase) o;

        return this.terms.equals(phrase.terms);
    }

    @Override
    public int hashCode() {
        return this.terms.hashCode();
    }

    @Override
    public String toString() {
        return printTerms(getTerms());
    }

    @Nonnull
    private Flux<PhraseExpansion> expansion(@Nonnull Resolver resolver) {
        return Flux.fromIterable(getTerms())
                .reduce(
                        Flux.just(new PhraseExpansion(resolver, getTerms().size())),
                        (expansions, term) -> expansions.flatMap(expansion -> expansion.expandTerm(term)))
                .flux()
                .flatMap(expansions -> expansions);
    }

    @Immutable
    private static final class PhraseExpansion {

        @Nonnull
        private final Predicate.Resolver resolver;

        @Nonnull
        private final Predicate predicate;

        @Nonnull
        private final PlainTerm[] terms;

        private final int index;

        PhraseExpansion(@Nonnull Predicate.Resolver resolver, int size) {
            this.resolver = resolver;
            this.predicate = Predicate.TRUE;
            this.terms = new PlainTerm[size];
            this.index = 0;
        }

        private PhraseExpansion(@Nonnull PhraseExpansion prev, @Nonnull Term.Expansion termExpansion) {
            this.resolver = prev.resolver.withKnowns(termExpansion.getKnowns());
            this.predicate = termExpansion.getUpdatePredicate().apply(prev.predicate);
            this.terms = prev.terms;
            this.index = prev.index + 1;
            this.terms[prev.index] = termExpansion.getExpanded();

        }

        @Nonnull
        Flux<PhraseExpansion> expandTerm(@Nonnull Term term) {
            return term
                    .expand(this.resolver)
                    .flux()
                    .map(termExpansion -> new PhraseExpansion(this, termExpansion));
        }

        @Nonnull
        Flux<Knowns> resolve() {
            return predicate().resolve(this.resolver);
        }

        @Nonnull
        Predicate definition(@Nonnull Variable local) {
            return this.predicate.and(local.definitionCall(this.terms));
        }

        @Nonnull
        private Predicate predicate() {
            return this.predicate.and(Predicate.call(Arrays.asList(this.terms)));
        }

    }

}

File: core/src/main/kotlin/org/predicode/predicator/Phrase.kt
package org.predicode.predicator

import org.predicode.predicator.grammar.TermPrinter
import org.predicode.predicator.grammar.printTerms
import reactor.core.publisher.Flux
import java.util.function.UnaryOperator

/**
 * A phrase consisting of other terms.
 *
 * @constructor constructs new phrase.
 *
 * @property terms list of terms this phrase consists of.
 */
class Phrase(val terms: List<Term>) : CompoundTerm() {

    /**
     * Constructs new phrase out of terms array.
     *
     * @param terms array of terms this phrase consists of.
     */
    constructor(vararg terms: Term) : this(terms.asList())

    override fun expand(resolver: PredicateResolver): Expansion? =
            expansion(resolver)?.let { expansion ->
                tempVariable("phrase expansion").let { tempVar ->
                    Expansion(
                            tempVar,
                            expansion.resolver.knowns,
                            UnaryOperator { expansion.definition(tempVar) and it })
                }
            }

    /**
     * Resolves this phrase as predicate.
     *
     * [expands][Term.expand] all of the phrase terms, then searches for corresponding [resolution rules][Rule]
     * and applies them.
     */
    fun resolve(resolver: PredicateResolver): Flux<Knowns> = try {
        expansion(resolver)?.resolve() ?: Flux.empty()
    } catch (e: Exception) {
        Flux.error(e)
    }

    override fun print(out: TermPrinter) {
        out.apply {
            startCompound()
            print(terms)
            endCompound()
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Phrase

        if (terms != other.terms) return false

        return true
    }

    override fun hashCode(): Int {
        return terms.hashCode()
    }

    override fun toString() = printTerms(terms)

    private fun expansion(resolver: PredicateResolver): PhraseExpansion? {

        val expansion = PhraseExpansion(resolver, terms.size)

        terms.forEach { term ->
            if (!expansion.expandTerm(term)) return null
        }

        return expansion
    }

    private class PhraseExpansion(var resolver: PredicateResolver, size: Int) {

        private var predicate: Predicate = True
        private val terms: Array<PlainTerm?> = arrayOfNulls(size)
        private var index = 0

        fun expandTerm(term: Term): Boolean =
                term.expand(resolver)
                        ?.let { (expanded, knowns, updatePredicate) ->
                            terms[index++] = expanded
                            predicate = updatePredicate.apply(predicate)
                            resolver = resolver.withKnowns(knowns)
                            true
                        }
                        ?: false

        fun resolve() = predicate().invoke(resolver)

        fun definition(variable: Variable): Predicate =
                resolver.knowns.declareLocal(variable) { local, knowns ->
                    resolver = resolver.withKnowns(knowns)
                    predicate and local.definitionOf(*expandedTerms()).applyRules()
                }

        private fun predicate() = predicate and RulePattern(*expandedTerms()).applyRules()

        @Suppress("UNCHECKED_CAST")
        private fun expandedTerms() = terms as Array<out PlainTerm>

    }

}
